harness/determined/common/api/request.py

```diff
@@ -106,11 +106,18 @@ def do_request(
     except requests.exceptions.RequestException as e:
         raise errors.BadRequestException(str(e))
 
+    def _get_message(r: requests.models.Response) -> str:
+        try:
+            return str(simplejson.loads(r.text).get("message"))
+        except Exception:
+            pass
+        return ""
+
     if r.status_code == 403:
         username = ""
         if auth is not None:
             username = auth.get_session_user()
-        raise errors.UnauthenticatedException(username=username)
+        raise errors.UnauthorizedException(username=username, message=_get_message(r))
```

Comment:
Hey Ryan,

I looked through this again and I don't see an instance in which you would arrive at a 403 if you are unauthenticated with the exception of:
- trying to login as a user with a valid token when the user has been deactivated in the database.
- In this instance, you are technically authenticated, but unauthorized.

What may have been confusing is the end-to-end test for "logout" - it is somewhat misleading. When we logout, it ends up falling back to DEFAULT_DETERMINED_USER and if that user happens to be deactivated, we return a 403. The `UnauthenticatedException` was thrown here to make it look like the user was unauthenticated.

I handled the above case a different way. I keep track of whether or not we fall back to the default user and mask the 403 as an UnauthenticatedException here: https://github.com/determined-ai/determined/commit/ef42dc0a4c5c20377ba8793fc82aefe141d37654#diff-7334dcb330f666bfb69eb6c226cf76cb9b0255050fe05db9d5d817a989284272R93

Reply:
I'd rename this to `ForbiddenException` given that 403 _means_ forbidden.  It's a matter of implemenation that, at the moment, it happens to always correspond to authenticated.

Right now what you have seems ok but some day when we support RBAC it will be nice for us to be picky about this now.

Reply:
Oh... I see, you are able to infer client-side what the 403 means.  Eh, I'd probably leave it as a ForbiddenException, and maybe modify the master to share some human-readable text that distinguishes the two states.

Reply:
Good call @rb-determined-ai . It is difficult to know how I should separate HTTP status and application level statuses. I think it makes sense to mirror the HTTP code in Python and then handle them differently based on context.

Done!